### DISCUSSION TITLE
Papis file renaming and metadata

### DISCUSSION THREAD

**Discussion by fab6** [papis/papis#836] [Q&A]
Hi,

I am further testing a few things in papis before I would switch from zotero.

For the folder and file renaming I would like to get a similar renaming like is
done in zotfile, e.g.
1) One author
    <1st_author_family_name>_<year>__<title-max_length_50_character>
1) Two authors
    <1st_author_family_name>_<2nd_author_family_name>_<year>__<title-max_length_50_character>
3) More than two authors
    <1st_author_family_name>_et_al_<year>__<title-max_length_50_character>

I saw these config options, but was not completely sure about the actual access
to e.g. just one author or even the different ways described above:   

add-folder-name = {doc[year]}_{doc[author]}_{doc[title]}
add-file-name = {doc[year]}_{doc[author]}_{doc[title]}
file-name = {doc[year]}_{doc[author]}_{doc[title]}

And somehow the underscores do not really work for me as well...

Further about the renaming, from your experience, does it make sense or is it at
all possible to add some
other tools to get good metadata. I am not sure how good papis works here
compared to zotero, but would like to have at least similar performance for
retrieving metadata from papers, books and report. There are some tools which
could extract some titles directly from the pdfs... maybe this is already
available anyways.

One example which I just tested was to add this document to papis and zotero:
https://doi.org/10.1038/s42254-023-00622-y

The result in papis was 
```
files:
- vinuesa-brunton-mckeon-natrevphys-2023.pdf
papis_id: 72db5ab78e5032a7431de7d2851b6e53
ref: '2024_04_29_15_47_47'
time-added: 2024-04-29-15:47:47
```
in Zotero I got the complete metadata
<img width="317" alt="image" src="https://github.com/papis/papis/assets/3047695/6f46ba80-9ea6-48a7-aa71-caa71fcafcb2">

For papis I used
```
papis add Vinuesa_Brunton_McKeon_NATREVPHYS_2023.pdf -d SciML/CFD
```

Maybe you have a hint, how to do this!?

Best regards
Fab

**Comment by alexfikl** [comment]
Hi! Thank you for your interest in `papis`. 
As you might have guessed, it's not as established and polished as Zotero, but we're getting better! :grin: 

> For the folder and file renaming I would like to get a similar renaming like is done in zotfile, e.g.
> 
> I saw these config options, but was not completely sure about the actual access to e.g. just one author or even the different ways described above:
> 
> add-folder-name = {doc[year]}_{doc[author]}_{doc[title]} 
> add-file-name = {doc[year]}_{doc[author]}_{doc[title]} 
> file-name = {doc[year]}_{doc[author]}_{doc[title]}

You're on the right track that those are the configuration options you should be modifying, but they probably need a lot of work to get them to the state you want in your examples. First off, these are simple Python formatting strings, see e.g.
https://docs.python.org/3/library/string.html#formatstrings

They're pretty limited in general. If you want to do fancy stuff like that, I would recommend trying out the `jinja2` formatter, which has a lot more control structures
https://jinja.palletsprojects.com/en/3.1.x/templates/#list-of-control-structures
As an example, my settings are simpler (don't support the "et al" part)
```
add-file-name = {{ doc.year }}-{{ doc.author_list|join("-", attribute="family") }}-{{ doc.title }}
add-folder-name = {{ doc.year }}-{{ doc.author_list|join("-", attribute="family") }}
```

Unfortunately switching to `jinja2` is a bit of a pain because you'll need to change ALL of the settings, but that will work much better once #711 is merged.
 
> And somehow the underscores do not really work for me as well...

This one should be easy to fix! If you're using the latest git version, you'll need to add the underscore `'_'` to the
[doc-paths-extra-chars](https://papis.readthedocs.io/en/latest/configuration.html#confval-doc-paths-extra-chars) configuration option. In older version this isn't supported at all, I'm afraid.

> Further about the renaming, from your experience, does it make sense or is it at all possible to add some other tools to get good metadata. I am not sure how good papis works here compared to zotero, but would like to have at least similar performance for retrieving metadata from papers, books and report. There are some tools which could extract some titles directly from the pdfs... maybe this is already available anyways.
> 
> ```
> papis add Vinuesa_Brunton_McKeon_NATREVPHYS_2023.pdf -d SciML/CFD
> ```
>
> Maybe you have a hint, how to do this!?

In general, it's very hard to extract data from just the PDF and papis is not very smart about it (yet!). If you have a DOI or an arXiv identifier or something like that, it's easier to directly tell it where to get the data from, e.g. for your document
```
papsi add --from doi '10.1038/s42254-023-00622-y' -d SciML/CFD Vinuesa_Brunton_McKeon_NATREVPHYS_2023.pdf
```
should do much better. This gets metadata from Crossref. We also support arXiv, dblp, isbn, pmid, zenodo, etc.

Hope that helps! One definitely nice thing about papis is that it's a lot smaller and easier to hack into than Zotero might be, so any help improving the workflow is very welcome!

**Reply by fab6** [reply]
Hi,
thank you for your quick reply!
It is fine, I don't need the most polished tool... cmdl is fine for me. I will look a bit at the format strings and see, how I can get it more like zotero.

About the metadata, thank you for the help here as well.

Would papis be able to read the doi itself from the paper? E.g. with some tools like `pdf2doi` or `pdf-rename` https://github.com/MicheleCotrufo/pdf-renamer ?

with pdf2doi it would be something like this
```
import pdf2doi
pdfF = "example.pdf"
results = pdf2doi.pdf2doi(pdfF)
doi = (results['identifier'])
print(doi)
```
Though, it seems when using the add from doi it does not take the available file which is already downloaded, but tries to get the file from the remote doi. Is it possible to get only the metadata and use the local copy from the pdf?
Thank you!
Fab

**Reply by alexfikl** [reply]
You can try `papis add --from pdf2doi 'somefile.pdf' somefile.pdf` to see if that works for you. However, it uses the very simple [python-doi](https://github.com/papis/python-doi) library, not `pdf2doi`, so it might not pick up the DOI as expected.

> Though, it seems when using the add from doi it does not take the available file which is already downloaded, but tries to get the file from the remote doi. Is it possible to get only the metadata and use the local copy from the pdf?

What version of papis are you using? In the git version, there's a `papis add --no-download-files` flag to stop it from downloading any files and only use the ones you give it on the command line. In papis 0.13, it shouldn't try to download a file if one was given on the command line, but I'm not sure if that worked correctly..

**Reply by fab6** [reply]
ok, I will check, thank you very much! And up to now I was using 0.13, but switch to the git version. Thank you... sorry, I just opened another discussion with a similar topic as I thought it would fit maybe better somewhere else.

**Comment by jghauser** [comment]
Hey @fab6,

Here's (the relevant part) of my `jinja2` config. Note that I keep this in my [`config.py`](https://papis.readthedocs.io/en/latest/configuration.html#python-configuration-file) as it makes escaping and such easier, but there is in principle no problem with doing this in the normal `config`. You'd still need to add stuff to handle more than one author (this snippet only ever uses the first author). Additionally, this snippet cuts filenames after 5 words. Also I'm not sure why I handle `surname`, I think this is a leftover from ages ago and Papis shouldn't ever set this rather than `family`.

```python
import papis.config

options = {
    "settings": {
        "add-file-name": '{% if doc["author_list"] %}'
        + '{{doc["author_list"][0].get("surname","") or doc["author_list"][0].get("family","")}}'
        + '{% elif doc["author"] %}'
        + '{{doc["author"].split()[0]}}'
        + '{% elif doc["editor"] %}'
        + '{{doc["editor"].split()[0]}}'
        + "{% endif %}"
        + "-"
        + '{% if doc["year"] %}'
        + '{{doc["year"]}}'
        + "-"
        + "{% endif %}"
        + '{% if doc["shorttitle"] %}'
        + '{{"-".join(doc["shorttitle"].split()[0:4])}}'
        + "{% else %}"
        + '{{"-".join(doc["title"].split()[0:4])}}'
        + "{% endif %}"
    }
}
papis.config.register_default_settings(options)

```

**Reply by fab6** [reply]
Hi,
great, thank you for your help! I will try to switch to the python config :-)
Thanks!

**Comment by fab6** [comment]
Hi,
now I am almost there, it took me a while :-)  

I am able to convert the whole zotero/zotfile data while keeping the subofolder structure.  I used a modified version of this julia repository `https://github.com/Klafyvel/ZoteroToPapis.jl`.

This was very important for me as I did not want to loose the current structure.

As I am using `emacs` I was also looking how to easily find/grep for pdfs. At the moment I am doing it with `consult` like:
```lisp
(defun fb/consult-zotfile-dir-fd-file ()
  (interactive)
  (let ((consult-fd-args "fd -e pdf"))
    (consult-fd fb/consult-find-zotero))
  )
(defun consult-papis-info-ripgrep ()
  "Search org-roam directory using consult-ripgrep. With live-preview."
  (interactive)
  (let ((consult-ripgrep-command "rg --null --ignore-case --type yaml --type-not pdf --line-buffered --color=always --max-columns=500 --no-heading --line-number . -e ARG OPTS"))
    (consult-ripgrep fb/papis)))

```
This works quite good, even with the roughly 8500 pdfs.

Now I am preparing a script for adding to my folder structure:

```python
import pdf2doi
import sys
import os
import json
import re

# ==============================================================================
def clean_filepath(filepath):
    cleaned_filepath = filepath.replace("ß", "ss")
    cleaned_filepath = cleaned_filepath.replace("ä", "ae")
    cleaned_filepath = cleaned_filepath.replace("ö", "oe")
    cleaned_filepath = cleaned_filepath.replace("ü", "ue")
    cleaned_filepath = cleaned_filepath.replace("©", "c")
    cleaned_filepath = re.sub(r"[^\x00-\x7F]", "_", cleaned_filepath)
    cleaned_filepath = cleaned_filepath.replace("-", "_")
    cleaned_filepath = cleaned_filepath.replace(" ", "_")
    cleaned_filepath = cleaned_filepath.replace("’", "_")
    cleaned_filepath = cleaned_filepath.replace(",", "_")
    cleaned_filepath = cleaned_filepath.replace(".", "_")
    return cleaned_filepath

# ==============================================================================
pdfF = sys.argv[1]
dstFolder = sys.argv[2]

# ==============================================================================
pdf2doi.config.set('save_identifier_metadata', False)

# ==============================================================================
print(100 * "=")
results = pdf2doi.pdf2doi(pdfF)
s = results["validation_info"]
json_acceptable_string = s.replace("'", "\"")
d = json.loads(json_acceptable_string)
title = d['title']
title = clean_filepath(title)
year = d['issued']['date-parts'][0][0]
author1 = d['author'][0]['family']

try:
    author2 = d['author'][1]['family']
except:
    author2 = None

try:
    author3 = d['author'][2]['family']
except:
    author3 = None

if author2 == None:
    part1 = author1

if author3 == None:
    part1 = author1 + "_" + author2

if author3 != None:
    part1 = author1 + "_et_al"
    
folderName = part1 + "__" + str(year) + "__" + title[0:50]

# ==============================================================================
addcmd = "papis add "

cmd = addcmd + "-d " + dstFolder  + " " + sys.argv[1] + " --folder-name " + folderName +  " --file-name " + folderName+".pdf "
print (100*"=")
print (100*"=")
print(cmd)
print (100*"=")
os.system(cmd)
```
This seems to work, except that the underscore and any capital characters are overwritten.
I expect that I can avoid somehow the use of `os.system`, but until now, I couldn't figure it out. Maybe you have an advice? It would be obviously great, if this can be integrated into `emacs/dired` but at the moment my `elisp` is not good enough for this.

Unfortunately, the python config with jinja2 is not clear to me yet. Especially, I still would like to handle the renaming in a more consitent way to the existing structure.

Thanks!
Fab

**Comment by fab6** [comment]
Hi,

just fyi
This is a slight adjustment to the above script which uses the `add` from the api and can be called by emacs from dired buffer:
```python
from papis.document import from_data
import papis.commands.add
import papis.commands.browse
import pdf2doi
import sys
import os
import json
import re

# ==============================================================================
def clean_filepath(filepath):
    cleaned_filepath = filepath.replace("ß", "ss")
    cleaned_filepath = cleaned_filepath.replace("ä", "ae")
    cleaned_filepath = cleaned_filepath.replace("ö", "oe")
    cleaned_filepath = cleaned_filepath.replace("ü", "ue")
    cleaned_filepath = cleaned_filepath.replace("©", "c")
    cleaned_filepath = re.sub(r"[^\x00-\x7F]", "_", cleaned_filepath)
    cleaned_filepath = cleaned_filepath.replace("-", "_")
    cleaned_filepath = cleaned_filepath.replace(" ", "_")
    cleaned_filepath = cleaned_filepath.replace("’", "_")
    cleaned_filepath = cleaned_filepath.replace(",", "_")
    cleaned_filepath = cleaned_filepath.replace(".", "_")
    return cleaned_filepath

# ==============================================================================
docFL = sys.argv[1:-1]
dstFolder = sys.argv[-1]

# ==============================================================================
pdf2doi.config.set('save_identifier_metadata', False)

# ==============================================================================
for i, docF in enumerate(docFL):
    print(100 * "=")
    results = pdf2doi.pdf2doi(docF)
    s = results["validation_info"]

    json_acceptable_string = s.replace("'", "\"")
    d = json.loads(json_acceptable_string)
    doi = d['DOI']
    title = d['title']
    title = clean_filepath(title)
    year = d['issued']['date-parts'][0][0]
    author1 = d['author'][0]['family']

    try:
        author2 = d['author'][1]['family']
    except:
        author2 = None
    try:
        author3 = d['author'][2]['family']
    except:
        author3 = None
    if author2 == None:
        part1 = author1
    if author3 == None:
        part1 = author1 + "_" + author2
    if author3 != None:
        part1 = author1 + "_et_al"

    folderName = part1 + "__" + str(year) + "__" + title
    filename, file_extension = os.path.splitext(docF)
    fileName = folderName + file_extension

    # ==============================================================================
    doiData = papis.api.doi_to_data(doi)
    print (100*"=")
    doiTitle = doiData['title']
    print("doiTitle: ", doiTitle)
    print("title: ", title)
    print (100*"-")
    print ("Adding doc: ", docF)
    if doiTitle == title:
        print (100*".")
        print (" - Adding based on DOI")
        papis.commands.add.run([docF],data=doiData, folder_name = folderName, file_name = fileName ,subfolder=dstFolder)
    else:
        print (100*".")
        print (" - Adding not based on DOI")
        docData = {"year": str(year), "author": author1 , "title": title}
        papis.commands.add.run([docF],data=docData, folder_name = folderName, file_name = fileName ,subfolder=dstFolder)
    print("DONE doc: ", docF)
print (100*".")
print("DONE")

```
and this can be used from within emacs:
```elisp
(defun papis-add-on-dired-files (&optional python-script)
  "Run a Python script asynchronously with marked files and `dired-dwim-target-directory`.
If no script is provided, use a default `copy_files.py` script."
  (interactive)
  (let* ((default-script
          "papis_api_addingDocument_dired.py"
          )  ;; Set default Python script path
         (script (or python-script default-script)) ;; Use provided script or fallback
         (files (mapcar #'shell-quote-argument (dired-get-marked-files))) ;; Get marked files
         (target-dir (shell-quote-argument (or (dired-dwim-target-directory) default-directory))) ;; Target dir
         (cmd (format "uv venv exec && python %s %s %s"
                      (shell-quote-argument (expand-file-name script))
                      (string-join files " ")
                      target-dir)))  
    (message "Running: %s" cmd) ;; Debugging output
    (async-shell-command cmd)))

```
